File: Python 3/0001_alt.py
import time
import sys
from fractions import gcd

# Euler1: Find the sum of all the multiples of 'a' or 'b' below 'limit'.


def euler1(a, b, limit):
    def lcm(a, b):
        """ Least common multiple """
        return (a * b) // gcd(a, b)

    def sci(n):
        """ Sum of consecutive integers """
        return n * (n + 1) / 2

    lcm_ac = lcm(a, b)
    limit -= 1
    return (
        a * (sci(limit // a)) +
        b * (sci(limit // b)) -
        lcm_ac * (sci(limit // lcm_ac))
    )


def euler1b(a, b, limit):
    s1 = set(range(a, limit, a))
    s2 = set(range(b, limit, b))
    ss = s1.union(s2)
    return sum(ss)


def euler1c(a, b, limit):
    return sum(
        [x for x in range(min(a, b), limit) if x % a == 0 or x % b == 0]
    )


def main():
    # take input from the user
    try:
        a, b, limit = map(int, sys.argv[1:4])
    except ValueError:
        a, b, limit = [3, 5, 1000000]
    print ("\nParams:", a, b, limit)

    for func in [euler1, euler1b, euler1c]:
        start_time = time.time()
        print (
            "\nResult {version_name}: {result}".format(
                version_name=func.__name__,
                result=func(a, b, limit)
            )
        )
        print ("--- {} seconds ---".format(time.time() - start_time))

if __name__ == "__main__":
    main()

File: Haskell/0001_alt.hs
module Main where

import qualified Data.List.Ordered as LO {-- data-ordlist package --}
import qualified Data.Set as Set
import Data.List
import System.Environment 
import Control.Exception
import Data.Time

{- Find the sum of all the multiples of 'a' or 'b' below 'limit' -}

euler1 :: (Integral a, Show a) => a -> a -> a -> IO ()
euler1 a b limit = print $ sumStep a limit + sumStep b limit - sumStep (lcm a b) limit
	where
		sumStep s n = s * sumOnetoN (n `div` s)
		sumOnetoN n = n * (n+1) `div` 2

euler1b :: (Integral a, Show a) => a -> a -> a -> IO ()
euler1b a b limit = print . sum . takeWhile (limit >=) $ LO.union [a,(a+a)..] [b,(b+b)..]

euler1c :: (Integral a, Show a) => a -> a -> a -> IO ()
euler1c a b limit = print . sum . Set.elems $ Set.union aa bb
	where
		aa = Set.fromList [a,(a+a)..limit]
		bb = Set.fromList [b,(b+b)..limit]

euler1d :: (Integral a, Show a) => a -> a -> a -> IO ()
euler1d a b limit = print . sum $ [x | x <- [3..limit], mod x a == 0 || mod x b == 0]

euler1e :: (Integral a, Show a) => a -> a -> a -> IO ()
euler1e a b limit = print . sum  $ union [a,(a+a)..limit] [b,(b+b)..limit]

main = run `catch` handler

run :: IO ()  
run = do
	args <- getArgs
	let params = map head $ transpose [argsi, defaults]
		where 
			argsi = [read x:: Integer| x <- take 3 args]
			defaults = [3, 5, 1000000]
	let (a:b:limit:_) = params
	
	putStr "\nParams: "; print params
	putStr "\nResult euler1: "; time $ euler1 a b (limit - 1)
	putStr "\nResult euler1b: "; time $ euler1b a b (limit - 1)
	putStr "\nResult euler1c: "; time $ euler1c a b (limit - 1)
	putStr "\nResult euler1d: "; time $ euler1d a b (limit - 1) 
	--putStr "\nResult euler1e: "; time $ euler1e a b (limit - 1) -- Very slow

handler :: SomeException -> IO ()  
handler e = putStrLn $ "Error. Usage: runhaskell 0001_alt.hs 3 5 1000"

time :: IO () -> IO ()
time a = do
    start <- getCurrentTime
    a
    stop   <- getCurrentTime
    let diff = diffUTCTime stop start
    putStr "Computation time: "
    print diff

